### PR TITLE
feat(background-media): add "Background media" UI component

### PR DESCRIPTION
**Pull request by proeung**
### Related Ticket(s)
https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/6129

DDS Consulting Issue: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/6394
HC JIRA Ticket: https://jsw.ibm.com/browse/HC-2014

### Description
- This PR adds the "Background media" UI component/pattern that designed to display an `image` or `video` with gradient options. For MVP we're looking to expose `image` for now, and later should be enhanced to support video background.
- This component is dependent of the L-shaped card group - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/6129.

### Design & Functional Specs
- https://github.com/carbon-design-system/carbon-for-ibm-dotcom-website/wiki/Background-media

68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3562653039633361363130323930306336643136643939312f31643833633730322d383565342d346165352d383636342d343364326330326633653735

### Changelog

**New Components/Props**

- `<dds-background-media>` – this new layout component is a variant/extension of the `<dds-image>`.
- `gradient-direction` - a prop/attribute (left-to-right (default) | top-to-bottom)
- `mobile-position` - a prop/attribute (top (default) | bottom)

**Removed**

- N/A

**Note**

- Need to figure out how to treat mobile image position where typically this should be defined within the layout component this `<dds-background-media>` is being used.

**Testing Instructions**
- Visit the `Background media` - `/?path=/story/components-background-media--default`
- Check the "Gradient direction" knobs
- Ensure that the component matches with the design specs for all breakpoints.

### CODE REVIEW

**Comment by shixiedesign** [comment]
Hey @proeung is this ready for design review?

**Comment by proeung** [comment]
@shixiedesign Yep! I just changed the PR status from `draft` to `ready for review`.

File: packages/styles/scss/components/background-media/_background-media.scss [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+@import 'carbon-components/scss/globals/scss/vendor/@carbon/elements/scss/themes/mixins';
+@import '../../globals/vars';
+@import '../../globals/imports';
+@import '../image/image';
+
+@mixin background-media {
+  :host(#{$dds-prefix}-background-media),
+  .#{$prefix}--background-media {
+    position: relative;
+  }
```
Comment (ariellalgilmore) [inline packages/styles/scss/components/background-media/_background-media.scss:17]:
```suggestion
  :host(#{$dds-prefix}-background-media) {
    position: relative;
  }
```

**Review by ariellalgilmore — APPROVED** [review]
LGTM!!!! Thank you @proeung

**Comment by shixiedesign** [comment]
Hi @proeung sorry for the wait! The only issue I noticed is **gradient top-to-bottom doesn't have the correct values.** The overlay should fade like this 

image

File: packages/styles/scss/components/background-media/_background-media.scss [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        to right,
+        $ui-background 25%,
+        rgba(255, 255, 255, 0) 75%
+      );
+    }
+  }
+
+  .#{$prefix}--background-media--gradient--top-to-bottom {
+    @include carbon--breakpoint('lg') {
+      background: linear-gradient(
```
Comment (kennylam) [inline packages/styles/scss/components/background-media/_background-media.scss:39]:
```suggestion
      background-image: linear-gradient(
```

File: packages/web-components/src/components/background-media/background-media.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+@customElement(`${ddsPrefix}-background-media`)
+class DDSBackgroundMedia extends DDSImage {
+  /**
+   * Returns a class-name based on the Gradient Direction type
+   */
+  protected _getGradientClass() {
+    return classMap({
+      [`${prefix}--background-media--gradient`]: true,
+      [`${prefix}--background-media--gradient--left-to-right`]: this.gradientDirection === GRADIENT_DIRECTION.LEFT_TO_RIGHT,
+      [`${prefix}--background-media--gradient--top-to-bottom`]: this.gradientDirection === GRADIENT_DIRECTION.TOP_TO_BOTTOM,
```
Comment (kennylam) [inline packages/web-components/src/components/background-media/background-media.ts:36]:
```suggestion
     [`${prefix}--background-media--gradient--${this.gradientDirection}`]: this.gradientDirection,
```

File: packages/web-components/src/components/background-media/background-media.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  /**
+   * Returns a class-name based on the Mobile Position type
+   */
+  protected _getMobilePositionClass() {
+    return classMap({
+      [`${prefix}--background-media--mobile-position`]: true,
+      [`${prefix}--background-media--mobile-position--top`]: this.mobilePosition === MOBILE_POSITION.TOP,
+      [`${prefix}--background-media--mobile-position--bottom`]: this.mobilePosition === MOBILE_POSITION.BOTTOM,
```
Comment (kennylam) [inline packages/web-components/src/components/background-media/background-media.ts:47]:
```suggestion
     [`${prefix}--background-media--mobile-position--${this.mobilePosition}`]: this.mobilePosition,
```

**Review by kennylam — COMMENTED** [review]
Thanks for this @proeung! Just a few small suggestions.

File: packages/styles/scss/components/background-media/_background-media.scss [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }
+  }
+
+  .#{$prefix}--background-media--gradient--top-to-bottom {
+    @include carbon--breakpoint('lg') {
+      background-image: linear-gradient(
+        to bottom,
+        $ui-background 0%,
+        rgba(255, 255, 255, 0) 86%
+      );
```
Comment (ariellalgilmore) [inline packages/styles/scss/components/background-media/_background-media.scss:43]: looking at shixie's feedback. I think these are the updates needed for the gradient. Let me know if you agree @proeung 

```suggestion
      background-image: linear-gradient(
        to bottom,
        $ui-background 25%,
        rgba(255, 255, 255, 0) 75%
      );
```

**Comment by proeung** [comment]
@shixiedesign I just pushed up a commit that adjusted the gradient value of the `top-to-bottom` direction to include a third opacity value as defined in the Sketch library. However, in doing so I have to expose a new prop as the current Carbon theme doesn't support light or dark `RGB` tokens.

Let me know if my approach makes sense. @ariellalgilmore and @annawen1 I'd like to get your input on this as well - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/6548/commits/21df9493384b92b903d32a2f3bd89e838ae1e7eb.

**Comment by ariellalgilmore** [comment]
> @shixiedesign I just pushed up a commit that adjusted the gradient value of the `top-to-bottom` direction to include a third opacity value as defined in the Sketch library. However, in doing so I have to expose a new prop as the current Carbon theme doesn't support light or dark `RGB` values unless there's a way to add to the theme tokens.
> 
> Let me know if my approach makes sense. @ariellalgilmore and @annawen1 I'd like to get your input on this as well - [21df949](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/commit/21df9493384b92b903d32a2f3bd89e838ae1e7eb).

hi @proeung! this workaround definitely works, but ideally we would want the component to change according to the theme and not need the adopter to set the attribute

**Comment by proeung** [comment]
@ariellalgilmore I agree. The workarounds implemented in this PR and in the [LS](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/styles/scss/components/leadspace/_leadspace.scss) are definitely not ideal, but it gets us to the `top-to-bottom` gradient value as defined in the design.

My ideal solution is to add `rgba()` tokens to the Carbon theme variables so that we can apply the theme based on what the adopter defines so that we can eliminate future tech debt and confusion. Also, another solution is to limit the `top-to-bottom` direction to two points as I had it previously and adjust the start and ending percentage.

**Review by shixiedesign — APPROVED** [review]
Visually this is good! I understand the current limitation with top-to-bottom gradient and will keep an eye on the Carbon issue. Thanks again Putra!

**Comment by shixiedesign** [comment]
Slightly concerned about Percy result here for Web components e2e test. Any idea why the masthead could be affected? These are seen in `dds-dotcom-shell | cdn | default` which is also not a story in web components storybook. If this is the CDN artifact adopters are getting, it's worth getting it right. But I could be missing something or just a percy error too. Lemme know. 

image
image

**Comment by jeffchew** [comment]
> Slightly concerned about Percy result here for Web components e2e test. Any idea why the masthead could be affected? These are seen in `dds-dotcom-shell | cdn | default` which is also not a story in web components storybook. If this is the CDN artifact adopters are getting, it's worth getting it right. But I could be missing something or just a percy error too. Lemme know.
> 

This is the link that this percy snapshot is built from:

https://webcomponents-codesandbox.s3-web.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/6548/dotcom-shell/cdn.html

This was just spun up pretty recently, but want to keep a close eye on this as the snapshot baseline stabilizes. 

cc: @ariellalgilmore

**Pull request merged**